### ISSUE TITLE
Clarify behaviour of Replace and Instr operator when first param is ommited

### ISSUE THREAD

**Issue by javihern98** [VTL-Community/VTL-Community#48]
Is this expression valid?:

```bash
replace("hello", _, "abc")
```

The reference manual does not make any reference to when the first parameter is omitted:

![image](https://github.com/user-attachments/assets/c7eaa929-87d0-444e-9f55-b181c49a08d9)

@linardian @vpinna80 @NicoLaval Any comment on this?

**Comment by javihern98** [comment]
Same behaviour on instr operator:

instr("testabcdef", _, 4, 3);

**Comment by vpinna80** [comment]
The second parameter cannot be omitted.

```
stringOperators:
    | REPLACE LPAREN expr COMMA param=expr ( COMMA optionalExpr)? RPAREN				                    # replaceAtom
    | INSTR LPAREN expr COMMA pattern=expr ( COMMA startParameter=optionalExpr)? (COMMA occurrenceParameter=optionalExpr)? RPAREN	            # instrAtom
;

stringOperatorsComponent:
    | REPLACE LPAREN exprComponent COMMA param=exprComponent ( COMMA optionalExprComponent)? RPAREN                                 # replaceAtomComponent
    | INSTR LPAREN exprComponent COMMA pattern=exprComponent ( COMMA startParameter=optionalExprComponent)? (COMMA occurrenceParameter=optionalExprComponent)? RPAREN    # instrAtomComponent
;
```

**Comment by javihern98** [comment]
@vpinna80 Yes but a null could be added, or empty parameter. The grammar currently allows these kind of expressions:

```sql
replace("hello", _, "abc")
instr("testabcdef", _, 4, 3);
instr("testabcdef", null, 4, 3);
```

That is why we request clarification on the usage of the first parameter and its nullability.

**Comment by vpinna80** [comment]
null and _ are not the same.
The replace's second parameter (or instr's third and fourth ones) is optional and, IMHO, non-nullable.
Since VTL doesn't have named parameters (yet), the _ is only given to mark the position - think of it as an overload of the same function with different parameter lists.

**Comment by linardian** [comment]
I agree with @vpinna80: _ and null are not equivalent and if a parameter can not be omitted (i.e. is mandatory) the "_" in the second parameter cannot be present. If we agree on this, the issue can be closed

**Comment by antonio-olleros** [comment]
I agree with the interpretation, but I think we should explicitly write that _ value is not allowed for non-optional parameters. It would even be better if this could be enforced at grammar level...

**Comment by linardian** [comment]
Agreed!

**Comment by javihern98** [comment]
Shall we close therefore this issue?

**Comment by antonio-olleros** [comment]
I think we should decide the action! It is either:

1. Enforce in the grammar (I imagine it is not possible)
2. Decide wher to specify in the manual

**Comment by linardian** [comment]
I have explicity stated in the manual (Additional constraints) that the second parameter cannot be omitted.
This is the new version of the replace operator:

String pattern replacement:   replace
Syntax		
replace (op , pattern1,  pattern2 )
Input parameters
op	          the operand
pattern1     the pattern (regular expression) to be replaced
pattern2     the replacing pattern
Examples of valid syntaxes		
replace(DS_1, "Hello", "Hi")
replace(DS_1, "Hello")
Semantics for scalar operations
Replaces all the occurrences of a specified string-pattern (pattern1) with another one (pattern2). If pattern2 is omitted then all occurrences of pattern1 are removed. For example:
replace("Hello world", "Hello", "Hi")	gives "Hi world"
replace("Hello world", "Hello")		gives " world"
replace ("Hello", "ello", "i")   		gives "Hi"
Input parameters type	
op ::	                	dataset   { measure<string>   _+ }
|   component<string>
|   string      
pattern1, pattern2 ::	component<string>
|  string
Result type
result ::	dataset   { measure<string>   _+ }
|   component<string>
|   string      
Additional constraints
The second operand (pattern1) cannot be omitted.
Behaviour 
As for the invocations at Data Set level, the operator has the behaviour of the “Operators applicable on one Scalar Value or Data Set or Data Set Component”, as for the invocations at Component or Scalar level, the operator has the behaviour of the “Operators applicable on more than two Scalar Values or Data Set Components”, (see the section “Typical behaviours of the ML Operators”).

**Comment by linardian** [comment]
The same for instr operator; I just changed operand --> parameter.
So for replace operator:
**Additional constraints.**
The second parameter (pattern1) cannot be omitted.

For instr operator:
**Additional constraints.**
The second parameter (pattern) cannot be omitted.
For operations at Data Set level, the input Data Set must have exactly one string type Measure.
